### PR TITLE
[expo-constants] Add new manifest2 field and make existing field optional

### PR DESCRIPTION
**Pull request by wschurman**
# Why

As we add a new manifest type, we need to reflect that there are now two types of manifests in the modules. This PR adds a new field to constants to express the new state of the world.

# How

This PR makes the manifest field nullable (reflecting the true type since it should have already been nullable for bare) and makes it null when the manifest is a new manifest. It also adds a new field `manifest2` that has the new manifest. I'm truly terrible at naming things so any naming advice would be appreciated. Unfortunately, to minimize breaking changes we can't call the new manifest "manifest".

This PR introduces one breaking change:
- Make the `manifest` field nullable.

This PR also changes the following (non-breaking):
- Reorganizing the Constants default export to be simpler (use getters instead of `Object.defineProperty`)

# Test Plan

`et cp --all`

### CODE REVIEW

File: packages/expo-constants/src/Constants.types.ts [hunk 1]
```diff
@@ -55,13 +92,21 @@ export interface AppManifest extends ExpoConfig {
     tool?: string;
     [key: string]: any;
   };
-  bundleUrl: string;
+  bundleUrl?: string;
   debuggerHost?: string;
   mainModuleName?: string;
   logUrl?: string;
-  [key: string]: any;
```
Comment (esamelson) [inline packages/expo-constants/src/Constants.types.ts:62]: This seems like a potential breaking change for developers who use the `AppManifest` type

**Review by esamelson — COMMENTED** [review]
Is it possible this will be a breaking change for any external developers who consume the current `AppManifest` type in their app logic and use fields that are now defined on `LegacyManifest` but not `NewManifest` (such as `revisionId`)? Or will typescript automatically infer that the object in question must have the `LegacyManifest` type?

**Comment by wschurman** [comment]
> Is it possible this will be a breaking change for any external developers who consume the current `AppManifest` type in their app logic and use fields that are now defined on `LegacyManifest` but not `NewManifest` (such as `revisionId`)? Or will typescript automatically infer that the object in question must have the `LegacyManifest` type?

Yep, though the change to setting the `Constants.manifest` to the new manifest already is an implicit breaking change whether or not we change the types. 

The only way to avoid this breaking change would be to a) put all the legacy manifest fields in the new manifest which would defeat the point, or b) create a translation layer from new manifest to legacy manifest so that the manifest we set `Constants.manifest` to is always a legacy manifest.

My preference is to go with the breaking change, providing a bunch of methods in the derived methods that are straightforward accessors for common needs (like we do in native code). I'm not familiar with our policy about breaking changes though. Alternatively we could go with one of the other ways mentioned above but then I think we should maybe reconsider the approach we're taking (creating a common interface between the two manifests) on all platforms if we're just going to be translating between them or adding all the fields into one.

One other option I thought of was to make `Constants.manifest` optional (breaking change, though [apparently already reality in bare apps](https://www.notion.so/expo/Manifests-and-app-config-13260052ccff4cf4a2402013491bc73d)) and adding another field called `Constants.theRealManifest` or something that has this union type. Because this still would be a breaking change though and wouldn't reflect the implicit underlying breaking change of the new data shape, I opted not to go this way. It seems preferable to have the types reflect the actual potential shapes of the data.

So overall the question is: are we okay making a breaking change to how we expose `Constants.manifest`? If not then do we prefer backwards translation of new manifests, deprecating/replacing `Constants.manifest` with something correctly typed, or putting all the old fields in the new manifest?

**Comment by esamelson** [comment]
> Yep, though the change to setting the `Constants.manifest` to the new manifest already is an implicit breaking change whether or not we change the types.

Fair enough 😄 thanks for the comprehensive answer; I agree with your preference in terms of moving forward with the breaking change and also guiding people away from directly depending on fields in `Constants.manifest`.

I guess my main concern is in the short/medium term for users who are not switching to EAS update and don't need to know or care about the new manifest format; it would be nice if developers for whom nothing is changing in terms of the actual shape of `Constants.manifest` didn't need to change anything in their app logic because of this type change. But if this is a critical and deliberate step in forcing developers to stop depending on these brittle fields, I think that's ok but would want to make sure it's marked as such in the changelog.

File: packages/expo-constants/src/Constants.ts [hunk 2]
```diff
@@ -106,11 +110,46 @@ const constants = {
     }
     return nativeConstants.linkingUri;
   },
+  get manifest() {
+    return getManifest();
+  },
+  get manifestDerivedMethods() {
+    if ('updateMetadata' in this.manifest) {
```
Comment (jkhales) [inline packages/expo-constants/src/Constants.ts:117]:
```suggestion
    if ('metadata' in this.manifest) {
```

File: packages/expo-constants/src/__tests__/Constants-test.ts [hunk 3]
```diff
@@ -159,3 +159,34 @@ describe(`manifest`, () => {
     });
   }
 });
+
+describe(`manifestDerivedMethods`, () => {
+  const legacyManifest = { id: '@jester/manifest', sdkVersion: '36.0.0' };
+  const newManifest = {
+    id: '6abaa89a-34e1-4c91-82fe-9139530dcf3f',
+    updateMetadata: {},
```
Comment (jkhales) [inline packages/expo-constants/src/__tests__/Constants-test.ts:167]:
```suggestion
    metadata: {},
```

File: packages/expo-constants/src/Constants.types.ts [hunk 4]
```diff
@@ -36,11 +36,48 @@ export interface WebManifest {
...
+export interface NewManifest {
+  id: string;
+  createdAt: string;
+  extras?: {
+    [k: string]: any;
+  };
+  runtimeVersion: string;
+  launchAsset: NewManifestAsset;
+  assets: NewManifestAsset[];
+  updateMetadata: object;
```
Comment (jkhales) [inline packages/expo-constants/src/Constants.types.ts:52]:
```suggestion
  metadata: object;
```

**Review by jkhales — COMMENTED** [review]
We are doing a rename 🙏 

https://github.com/expo/expo/pull/12818#issuecomment-833163174

File: packages/expo-auth-session/CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- [updates] Add new manifest types to various non-native module code. ([#12817](https://github.com/expo/expo/pull/12817) by [@wschurman](https://github.com/wschurman))
```
Comment (bbarthec) [inline packages/expo-auth-session/CHANGELOG.md:9]:
```suggestion
- Add new manifest types to various non-native module code. ([#12817](https://github.com/expo/expo/pull/12817) by [@wschurman](https://github.com/wschurman))
```

File: packages/expo-constants/CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- [updates] Add new manifest types to various non-native module code. ([#12817](https://github.com/expo/expo/pull/12817) by [@wschurman](https://github.com/wschurman))
```
Comment (bbarthec) [inline packages/expo-constants/CHANGELOG.md:9]:
```suggestion
- Add new manifest types to various non-native module code. ([#12817](https://github.com/expo/expo/pull/12817) by [@wschurman](https://github.com/wschurman))
```

File: packages/expo-linking/CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- [updates] Add new manifest types to various non-native module code. ([#12817](https://github.com/expo/expo/pull/12817) by [@wschurman](https://github.com/wschurman))
```
Comment (bbarthec) [inline packages/expo-linking/CHANGELOG.md:9]:
```suggestion
- Add new manifest types to various non-native module code. ([#12817](https://github.com/expo/expo/pull/12817) by [@wschurman](https://github.com/wschurman))
```

**Review by ide — COMMENTED** [review]
Since the names of these types end up being part of the public API, I suggest:

* NewManifest -> Manifest
* LegacyManifest -> ClassicManifest
* AppManifest -> IndeterminateManifest or AnyManifest

But perhaps this is unnecessary: we're gradually paring down Constants and moving the fields to more more specific libraries (maybe one day expo-constants will go away) -- namely, expo-updates. In short, should we just set `Constants.manifest` to undefined if the app is using a modern manifest and instead have the app read from `expo-updates` instead? This way the type of `Constants.manifest` would be `manifest?: ClassicManifest`.

**Comment by wschurman** [comment]
> In short, should we just set `Constants.manifest` to undefined if the app is using a modern manifest and instead have the app read from `expo-updates` instead? This way the type of `Constants.manifest` would be `manifest?: ClassicManifest`.

I don't feel strongly one way or the other, and see the pros and cons of both. I think not needing to dig into two different libraries at each callsite is a nice feature of having the manifest in one place (and be an `or` type), but I remember also seeing a comment somewhere that the manifest should be moved to expo-updates. Maybe the right call is to remove `Constants.manifest` and exclusively move it to `expo-updates` and still have it be an `or` type there?

**Comment by ide** [comment]
In the end state I think it'd be nice to have a field named `manifest` with the type declaration that defines only a modern manifest. That will be the common case for developers in the future.

How could we forge a path there (that might involve deprecation steps over time, but ideally minimizing thrash from breaking changes)?

**Comment by wschurman** [comment]
> In the end state I think it'd be nice to have a field named manifest with the type declaration that defines only a modern manifest. That will be the common case for developers in the future.

This seems like it will be many many years in the future since the plan it to keep supporting legacy updates and manifests indefinitely or at least for a long time. I agree on the end state being just a new manifest but since the interim state is so long it probably makes sense to design something that optimizes for it instead of the end state.

> How could we forge a path there (that might involve deprecation steps over time, but ideally minimizing thrash from breaking changes)?

I'm not too familiar with our policy on breaking changes, but if we're trying to avoid breaking changes over a single SDK version then here's what I propose:
1. Deprecate `Constants.manifest`, make it nullable over a single SDK version upgrade (it already sounds like it technically is nullable). Set it to null for new manifests. Add a console.warn or something when the old manifest field is accessed.
2. Add another field, `Constants.theRealManifest` (jk, but something like that) or a field in `expo-updates` that is the `or` type as is defined in this PR.
3. In all our first party libraries use that new field.
4. Wait a few SDK versions, delete the old `Constants.manifest` field.
5. Gradually phase out legacy updates over many years.
6. When it comes time to deprecate legacy updates, console.warn when accessing legacy manifest fields.
7. Wait a few SDK versions, make `Constants.theRealManifest` just the new manifest type.

@ide - How does that sound?

**Comment by ide** [comment]
Developers can very quickly reach the end state just by using an update service with modern manifests. So even though we'll need to support classic manifests for a longer time, we want developers who are using modern manifests to have a nice API and type declarations without the classic manifests in the way.

File: packages/expo-constants/src/Constants.ts (hunk 2 above)
Comment (wschurman) [reply packages/expo-constants/src/Constants.ts]: I'll do this all in one follow-up commit since there are a lot of other places that assume `updateMetadata`.

**Comment by wschurman** [comment]
Updated this PR to have fewer breaking changes, though I think making the field nullable is itself a breaking change. The key is that it technically already was nullable since it's nullable in bare.

The `expo-dev-launcher` changes are just because I ran `et cp --all` and these files were uncommitted so I didn't add a changelog entry.

File: packages/expo-constants/src/Constants.ts [hunk 8]
```diff
@@ -106,11 +108,48 @@ const constants = {
...
 
-function getManifest(suppressWarning = false) {
-  if (!manifest) {
-    const invalidManifestType = manifest === null ? 'null' : 'undefined';
+function isAppManifest(manifest: AppManifest | Manifest): manifest is AppManifest {
+  return !isManifest(manifest);
+}
+
+function isManifest(manifest: AppManifest | Manifest): manifest is Manifest {
+  return 'updateMetadata' in manifest;
```
Comment (ide) [inline packages/expo-constants/src/Constants.ts:147]:
```suggestion
  return 'metadata' in manifest;
```

File: packages/expo-constants/src/Constants.ts [hunk 9]
```diff
@@ -106,11 +108,48 @@ const constants = {
...
+    }
+    return maybeManifest;
+  },
+  get manifest2(): Manifest | null {
+    const maybeManifest = getManifest();
+    if (!maybeManifest || !isManifest(maybeManifest)) {
+      return null;
+    }
+    return maybeManifest;
+  },
```
Comment (ide) [inline packages/expo-constants/src/Constants.ts:124]: Docblocks on these two methods would help a lot since "manifest2" otherwise doesn't make sense to someone who's unaware of the change from classic updates to modern updates.

File: packages/expo-updates/src/Updates.types.ts [hunk 10]
```diff
@@ -7,7 +7,7 @@ export enum UpdateEventType {
 }
 
 // TODO(eric): move source of truth for manifest type to this module
-export type Manifest = typeof Constants.manifest;
+export type Manifest = typeof Constants.manifest | typeof Constants.manifest2;
```
Comment (ide) [inline packages/expo-updates/src/Updates.types.ts:10]: What do you think of something like this?
```suggestion
export type ClassicManifest = typeof Constants.manifest;
export type Manifest = ClassicManifest | typeof Constants.manifest2;
// modern manifest type is intentionally not exported, since the plan is to call it just "Manifest" in the future
```

and then later just:
```
export type Manifest = typeof Constants.manifest2;
export type ClassicManifest = typeof Constants.manifest;
```

File: packages/expo-constants/src/Constants.types.ts [hunk 11]
```diff
@@ -104,7 +123,16 @@ export interface NativeConstants {
   linkingUri: string;
   nativeAppVersion: string | null;
   nativeBuildVersion: string | null;
-  manifest: AppManifest;
+  /**
+   * Classic manifest for Expo apps.
```
Comment (ide) [inline packages/expo-constants/src/Constants.types.ts:127]:
```suggestion
   * Classic manifest for Expo apps using classic updates.
```

File: packages/expo-constants/src/Constants.types.ts [hunk 12]
```diff
@@ -104,7 +123,16 @@ export interface NativeConstants {
...
+   * Returns `null` in bare workflow and when `manifest2` is non-null.
+   */
+  manifest: AppManifest | null;
+  /**
+   * New manifest for Expo apps using expo-updates.
```
Comment (ide) [inline packages/expo-constants/src/Constants.types.ts:132]:
```suggestion
   * New manifest for Expo apps using modern Expo Updates.
```

**Review by ide — APPROVED** [review]

**Comment by wschurman** [comment]
Failing test is also broken in master but wasn't run due to "Using incremental checks" (blame https://github.com/expo/expo/commit/207019eb8b76f3baf3ae0765773d26254d072f5b).

**Comment by SimenB** [comment]
Hiya 👋 Sorry to comment on an old PR, but is there anyplace I can read about the background for this change? And a migration path? This change is not mentioned in the [blog post](https://blog.expo.io/expo-sdk-42-579aee2348b6) and I see no linked issues here. There is talk about a "modern update" - what's that? Is it EAS, or something for all users? If (when) we move to EAS how can I access e.g. `debuggerHost` to make requests to a locally running server?

Tl;dr: is it safe for me using managed workflow to just do `manifest!.blah`, and what would be the change (either I as a dev or Expo as a platform) would make for that _not_ to be safe?

**Comment by wschurman** [comment]
> Hiya 👋 Sorry to comment on an old PR, but is there anyplace I can read about the background for this change? And a migration path? This change is not mentioned in the [blog post](https://blog.expo.io/expo-sdk-42-579aee2348b6) and I see no linked issues here. There is talk about a "modern update" - what's that? Is it EAS, or something for all users?

This change is part of a series of ongoing changes to add support for Expo Updates to our tools and services. Expo Updates is an upcoming feature we're developing as part of EAS.

> If (when) we move to EAS how can I access e.g. `debuggerHost` to make requests to a locally running server?

The new manifest format is https://docs.expo.io/technical-specs/expo-updates-0/. We're still ironing out the details of where specific fields like `debuggerHost` will live in the new manifest format. A recent PR that made some progress in this area was https://github.com/expo/expo-cli/pull/3606. When we launch Expo Updates into [feature preview](https://docs.expo.io/feature-preview/) we will document everything as well as a detailed migration path.

> Tl;dr: is it safe for me using managed workflow to just do `manifest!.blah`, and what would be the change (either I as a dev or Expo as a platform) would make for that _not_ to be safe?

It will continue to be safe to do `manifest!.blah` using managed workflow. When you as a developer decide to migrate over to Expo Updates on EAS, Expo will provide a detailed migration guide and that will include instructions to change these callsites to `manifest2!.blah` (or something like it).

**Comment by ide** [comment]
We are thinking about a standard specification for universal updates. This is a stable specification for how to deliver and run updates in Expo clients (like Expo Go, custom development clients, and standalone apps made with Expo) similar to how HTML is a specification for how to deliver and run pages in web clients. The first draft of what we're calling Expo Updates is on [the docs](https://docs.expo.io/technical-specs/expo-updates-0/) (still in flux).

Expo is designed to be independent from EAS, like how Chrome is independent from Google Cloud though you certainly can use Google Cloud to build and host websites. Just like you can host websites on any compatible web server, we're designing Expo Updates so you can host updates on any compatible update server. We're also thinking about how EAS could provide an update server for people who are looking for a service run by the Expo team that integrates with other parts of EAS like your team accounts and builds.

At some point you'll have the option to switch Expo CLI to use the new update format described by the Expo Updates spec. When you enable this, Expo CLI will serve update manifests in the new format and you'll have to make some small changes to your application code and we'll provide guidance like @wschurman said.

Additionally, you'll be able to export updates in the new format and host them on your own server (we think we can provide a simple example implementation of an Expo Updates server written in Node and you can also read the Expo Updates spec and write your own). A new update service as part of EAS ("EAS Update", like a counterpart to "EAS Build"?) would support the Expo Updates specification and the classic update service (`expo publish`) would support classic updates.

**Comment by SimenB** [comment]
Wonderful, thank you for the detailed responses! I read `Expo Updates` as `expo-updates` which we already use, but if it's a new service then my worries are unfounded 🙂

Very much looking forward to following the development of EAS (and related services) into the future! 🚀

**Comment by ide** [comment]
**Expo Updates** is the name of the new standard specification for universal updates that has a v0 draft [here](https://docs.expo.io/technical-specs/expo-updates-0/). **Classic updates** is the existing update format that you use with `expo publish` and `expo export`. In the future, you will be able to use Expo CLI to export updates in the new Expo Updates format and host them on a server that implements the server-side part of the Expo Updates spec.

**expo-updates** is the name of the client library that is both a classic updates client (mature) and an Expo Updates client (not very mature yet).

**EAS Update** is a potential future service built by the Expo team that would be an Expo Updates server as part of EAS. We see some opportunities to make updates even faster and more globally available with a new service.

We're designing Expo Updates, and Expo in general, to be independent from EAS so that developers can choose whether to use EAS Update or to write & run their own server based on the merits of each. (It's sort of like how Google Cloud Platform lets you host websites that people can visit with Google Chrome, but Chrome does not grant special privileges to sites hosted by GCP -- GCP needs to attract developers on its own merit.)
